study1/week3/evaluate-the-bracket-pairs-of-a-string.java

```diff
@@ -0,0 +1,26 @@
+// https://leetcode.com/problems/evaluate-the-bracket-pairs-of-a-string
+// Time Complexity: O(n)
+// Space Complexity: O(n), because of hashmap
+class Solution {
+    public String evaluate(String s, List<List<String>> knowledge) {
+        String result = "";
```

Comment:
When building strings with concatenation, the majority of the cases it's best to use [StringBuilder](https://docs.oracle.com/javase/8/docs/api/java/lang/StringBuilder.html).

The reason is the following. Java Strings are immutable, meaning that you cannot edit, append or remove characters arbitrary from String types. So when you're concatenating 'a' to 'b' what's happening internally is that you allocate a space containing enough space for both a, and b. then you copy into that space a and b. 

So lets say there's no parenthesis in an input and it has 10 characters.
The amount of copy happening in each iteration of line 20, is
1st iter: 1, 2nd iter : 2( 1+1), 3rd iter : 3(2+1) and so on. 
So the time complexity of this function because due to using String concatenation, is `O(n^2)`, same goes for space complexity.